File: src-tauri/src/ssh.rs
// SSH connection spawning. Wraps portable-pty CommandBuilder to launch ssh
// (optionally via sshpass) using the shared PTY plumbing in pty.rs.

use anyhow::{anyhow, Result};
use portable_pty::CommandBuilder;
use tauri::ipc::Channel;

use crate::hosts;
use crate::pty::{spawn_with_command, PtyEvent};

fn ssh_args(host: &hosts::HostMeta) -> Vec<String> {
    let mut args: Vec<String> = vec![
        "-t".into(),
        "-o".into(),
        "ServerAliveInterval=30".into(),
        "-p".into(),
        host.port.to_string(),
    ];
    match host.auth.as_str() {
        "key" => {
            if let Some(path) = host.identity_path.as_ref().filter(|s| !s.is_empty()) {
                args.push("-i".into());
                args.push(path.clone());
                args.push("-o".into());
                args.push("IdentitiesOnly=yes".into());
            }
            args.push("-o".into());
            args.push("PreferredAuthentications=publickey".into());
        }
        "password" => {
            // sshpass feeds "password:" prompt — must skip pubkey auth or ssh
            // hangs on "Enter passphrase for key" prompt that sshpass doesn't match.
            args.push("-o".into());
            args.push("PubkeyAuthentication=no".into());
            args.push("-o".into());
            args.push("PreferredAuthentications=password".into());
        }
        "agent" => {
            args.push("-o".into());
            args.push("PreferredAuthentications=publickey".into());
            args.push("-o".into());
            args.push("IdentityAgent=$SSH_AUTH_SOCK".into());
        }
        _ => {}
    }
    args.push(format!("{}@{}", host.user, host.host));
    args
}

fn build_command(
    host: &hosts::HostMeta,
    password: Option<&str>,
    debug: &mut String,
) -> Result<CommandBuilder> {
    let args = ssh_args(host);
    let mut cmd;
    let mut display: Vec<String> = Vec::new();
    match host.auth.as_str() {
        "password" => {
            let pw = password.ok_or_else(|| anyhow!("password required but missing"))?;
            cmd = CommandBuilder::new("sshpass");
            cmd.arg("-p");
            cmd.arg(pw);
            cmd.arg("ssh");
            display.push("sshpass".into());
            display.push("-p".into());
            display.push("***".into());
            display.push("ssh".into());
            for a in &args {
                cmd.arg(a);
                display.push(a.clone());
            }
        }
        _ => {
            cmd = CommandBuilder::new("ssh");
            display.push("ssh".into());
            for a in &args {
                cmd.arg(a);
                display.push(a.clone());
            }
        }
    }

    if let Ok(home) = std::env::var("HOME") {
        cmd.env("HOME", &home);
        cmd.cwd(home);
    }
    if let Ok(v) = std::env::var("PATH") {
        cmd.env("PATH", v);
    }
    if let Ok(v) = std::env::var("USER") {
        cmd.env("USER", v);
    }
    if let Ok(v) = std::env::var("LOGNAME") {
        cmd.env("LOGNAME", v);
    }
    if let Ok(v) = std::env::var("SSH_AUTH_SOCK") {
        cmd.env("SSH_AUTH_SOCK", v);
    }
    if let Ok(v) = std::env::var("LANG") {
        cmd.env("LANG", v);
    }
    cmd.env("TERM", "xterm-256color");
    cmd.env("COLORTERM", "truecolor");
    cmd.env("MTERMINAL", "1");

    *debug = display.join(" ");
    Ok(cmd)
}

#[tauri::command]
pub fn ssh_spawn(
    events: Channel<PtyEvent>,
    rows: u16,
    cols: u16,
    host_id: String,
) -> Result<u32, String> {
    let host = hosts::get_host(&host_id).map_err(|e| e.to_string())?;

    let password = if host.auth == "password" {
        if host.save_password {
            let pw = hosts::host_get_password(host.id.clone())?;
            Some(pw.ok_or_else(|| {
                "no saved password for this host — edit host or unlock vault".to_string()
            })?)
        } else {
            return Err(
                "password auth without saved password is not supported — save the password or use key auth".into(),
            );
        }
    } else {
        None
    };

    let mut debug = String::new();
    let cmd = build_command(&host, password.as_deref(), &mut debug).map_err(|e| e.to_string())?;

    let banner = format!("\x1b[2m[exec] {}\x1b[0m\r\n", debug);
    let _ = events.send(PtyEvent::Data(banner));

    let id = spawn_with_command(events, rows, cols, cmd).map_err(|e| e.to_string())?;
    let _ = hosts::touch_last_used(&host_id);
    Ok(id)
}

File: electron/main/ssh.ts
import { ipcMain } from 'electron'
import fs from 'node:fs'
import path from 'node:path'
import os from 'node:os'
import * as nodePty from 'node-pty'
import {
  RingBuffer,
  SESSIONS,
  nextId,
  getMainWindow,
  type PtySession,
} from './sessions'
import { getHost, getHostPassword, touchLastUsed } from './hosts'

interface SshSpawnArgs {
  rows: number
  cols: number
  hostId: string
}

export interface HostMeta {
  id: string
  name: string
  host: string
  port: number
  user: string
  auth: string
  identityPath?: string
  savePassword?: boolean
}

export function whichOnPath(prog: string): string | null {
  const PATH = process.env.PATH
  if (!PATH) return null
  const sep = process.platform === 'win32' ? ';' : ':'
  const isWin = process.platform === 'win32'
  const exts = isWin
    ? (process.env.PATHEXT || '.EXE;.CMD;.BAT')
        .split(';')
        .filter((s) => s.length > 0)
    : ['']
  for (const dir of PATH.split(sep)) {
    if (!dir) continue
    const base = path.join(dir, prog)
    for (const ext of exts) {
      const candidate = ext ? base + ext : base
      try {
        const st = fs.statSync(candidate)
        if (st.isFile()) return candidate
      } catch {}
    }
  }
  return null
}

export function sshArgs(host: HostMeta): string[] {
  const args: string[] = [
    '-t',
    '-o',
    'ServerAliveInterval=30',
    '-p',
    String(host.port),
  ]
  switch (host.auth) {
    case 'key':
      if (host.identityPath && host.identityPath.length > 0) {
        args.push('-i', host.identityPath, '-o', 'IdentitiesOnly=yes')
      }
      args.push('-o', 'PreferredAuthentications=publickey')
      break
    case 'password':
      args.push('-o', 'PubkeyAuthentication=no')
      args.push('-o', 'PreferredAuthentications=password')
      break
    case 'agent':
      args.push('-o', 'PreferredAuthentications=publickey')
      args.push('-o', 'IdentityAgent=$SSH_AUTH_SOCK')
      break
  }
  args.push(`${host.user}@${host.host}`)
  return args
}

export function buildEnv(): Record<string, string> {
  const env: Record<string, string> = {}
  const home = process.env.HOME || process.env.USERPROFILE || os.homedir()
  if (home) env.HOME = home
  for (const k of ['PATH', 'USER', 'LOGNAME', 'SSH_AUTH_SOCK', 'LANG']) {
    const v = process.env[k]
    if (typeof v === 'string') env[k] = v
  }
  env.TERM = 'xterm-256color'
  env.COLORTERM = 'truecolor'
  env.MTERMINAL = '1'
  return env
}

function spawnRaw(opts: {
  rows: number
  cols: number
  command: string
  args: string[]
  env: Record<string, string>
}): number {
  const cwd =
    process.env.HOME || process.env.USERPROFILE || os.homedir() || process.cwd()
  const ptyProc = nodePty.spawn(opts.command, opts.args, {
    name: 'xterm-256color',
    cols: Math.max(1, opts.cols | 0),
    rows: Math.max(1, opts.rows | 0),
    cwd,
    env: opts.env,
  })

  const id = nextId()
  const ring = new RingBuffer()
  const session: PtySession = {
    id,
    pid: ptyProc.pid,
    pty: ptyProc,
    ringBuffer: ring,
    lastActivityMs: Date.now(),
    shell: opts.command,
  }
  SESSIONS.set(id, session)

  const channel = 'pty:event:' + id
  ptyProc.onData((chunk) => {
    const buf = Buffer.from(chunk, 'utf8')
    ring.push(buf)
    session.lastActivityMs = Date.now()
    const win = getMainWindow()
    if (win && !win.isDestroyed()) {
      win.webContents.send(channel, { kind: 'data', value: chunk })
    }
  })
  ptyProc.onExit(() => {
    const win = getMainWindow()
    if (win && !win.isDestroyed()) {
      win.webContents.send(channel, { kind: 'exit' })
    }
    SESSIONS.delete(id)
  })

  return id
}

function emitBanner(id: number, banner: string): void {
  const win = getMainWindow()
  if (win && !win.isDestroyed()) {
    win.webContents.send('pty:event:' + id, { kind: 'data', value: banner })
  }
}

export function registerSshHandlers(): void {
  ipcMain.handle('ssh:spawn', async (_e, args: SshSpawnArgs) => {
    const host = getHost(args.hostId) as HostMeta | null
    if (!host) throw new Error('host not found')

    let password: string | null = null
    if (host.auth === 'password') {
      if (!host.savePassword) {
        throw new Error(
          'password auth without saved password is not supported — save the password or use key auth'
        )
      }
      try {
        password = getHostPassword(host.id)
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err)
        if (/vault.*lock/i.test(msg)) throw new Error('vault locked')
        throw err
      }
      if (!password) throw new Error('no saved password for host')
    }

    const baseArgs = sshArgs(host)
    let command: string
    let argv: string[]
    let display: string[]
    if (host.auth === 'password') {
      const sshpassPath = whichOnPath('sshpass')
      if (!sshpassPath) throw new Error('sshpass missing on PATH')
      command = 'sshpass'
      argv = ['-p', password as string, 'ssh', ...baseArgs]
      display = ['sshpass', '-p', '***', 'ssh', ...baseArgs]
    } else {
      command = 'ssh'
      argv = baseArgs
      display = ['ssh', ...baseArgs]
    }

    const env = buildEnv()
    const id = spawnRaw({
      rows: args.rows,
      cols: args.cols,
      command,
      args: argv,
      env,
    })
    emitBanner(id, `\x1b[2m[exec] ${display.join(' ')}\x1b[0m\r\n`)

    try {
      await touchLastUsed(args.hostId)
    } catch {}
    return id
  })
}
